leanplum_data_export/export.py

```diff
@@ -0,0 +1,133 @@
+        response = requests.get(export_init_url)
+        response.raise_for_status()
+
+        # Will hard fail if not present
+        return response.json()["response"][0]["jobId"]
+
+    def get_files(self, job_id, sleep_time=DEFAULT_SLEEP_SECONDS):
+        export_retrieve_url = (f"http://www.leanplum.com/api?"
+                               f"appId={self.app_id}"
+                               f"&clientKey={self.client_key}"
+                               f"&apiVersion=1.0.6"
+                               f"&action=getExportResults"
+                               f"&jobId={job_id}")
+
+        logging.info("Export URL: " + export_retrieve_url)
+        loading = True
+        while(loading):
+            response = requests.get(export_retrieve_url)
+            response.raise_for_status()
+
+            state = response.json()['response'][0]['state']
+            if (state == LeanplumExporter.FINISHED_STATE):
+                logging.info("Export Ready")
+                loading = False
+            else:
+                logging.info("Waiting for export to finish...")
+                time.sleep(sleep_time)
+
+        return response.json()['response'][0]['files']
+
+    def save_files(self, file_uris, bucket_name, prefix, date, export_format):
+        client = storage.Client()
+        bucket = client.get_bucket(bucket_name)
+        datatypes = set()
+
+        if prefix:
+            prefix = self.add_slash_if_not_present(prefix) + date
+        else:
+            prefix = date
+
+        for uri in file_uris:
+            logging.info(f"Retrieving URI {uri}")
+
+            parsed = self.filename_re.fullmatch(uri)
+            if parsed is None:
+                raise Exception(f"Expected uri matching {LeanplumExporter.FILENAME_RE}, but got {uri}")
+
+            datatype, index = parsed.group(1), parsed.group(2)
+            local_filename = f"{datatype}/{index}.{export_format}"
+            datatypes |= set([datatype])
+
+            f = Path(local_filename)
+            base_dir = Path(datatype)
+            base_dir.mkdir(parents=True, exist_ok=True)
+
+            with requests.get(uri, stream=True) as r:
+                r.raise_for_status()
+                with f.open("wb") as opened:
+                    for chunk in r.iter_content(chunk_size=8192):
+                        if chunk:
+                            opened.write(chunk)
+
+            logging.info(f"Uploading to gs://{bucket_name}/{prefix}/{local_filename}")
+            blob = bucket.blob(f"{prefix}/{local_filename}")
+            blob.upload_from_filename(local_filename)
+
+            f.unlink()
+            base_dir.rmdir()
+
+        return datatypes
+
+    def create_external_tables(self, bucket_name, prefix, date, tables, dataset):
+        gcs_loc = f"gs://{bucket_name}/{prefix}/{date}"
+
+        client = bigquery.Client()
+
+        dataset_ref = client.dataset(dataset)
+
+        for t in tables:
```

Comment:
Leanplum API returns files of the form:
`https://leanplum_export.storage.googleapis.com/export-$JOB_ID-$FILE_TYPE-$FILE_NO_OF_TYPE`

Where `$FILE_TYPE` is something like `outputsessions` and `outputexperiments`. They are all joinable to get e.g. events.

From there, we create tables of the form `$FILETYPE_$DATE`, which BQ handles well by default.